### ISSUE TITLE
lld/mac emits too many STABS entries when doing ICF

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#54783]
(I have a patch for this I'll upload in a bit. Just want to file a bug so I have a place to add some more detail.)

Consider this example:

```
% cat bar.cc
__attribute__((noinline)) inline int quux(int x) { return x; }
__attribute__((noinline)) inline int shromp(int x) { return x; }
__attribute__((noinline)) inline int freep(int x) { return x; }
int g(int i) {
  return quux(i) + shromp(i) + freep(i);
}

% cat baz.cc
__attribute__((noinline)) inline int foo(int x) { return x; }
__attribute__((noinline)) inline int bar(int x) { return x; }
__attribute__((noinline)) inline int baz(int x) { return x; }
int g(int);
int main(int argc, char* argv[]) {
  return foo(argc) + bar(argc) + baz(argc) + g(argc);
}

% clang -O2 bar.cc baz.cc -c -g
```

Let's link it with ld64:

```
% out/gn/bin/clang -O2 bar.o baz.o -Wl,-verbose_deduplicate
deduplicate the following 2 functions (8 bytes apiece):
    __Z5freepi
    __Z3bazi
deduplicate the following 4 functions (16 bytes apiece):
    __Z4quuxi
    __Z6shrompi
    __Z3fooi
    __Z3bari
deduplication saved 56 bytes of __text
% dsymutil -s a.out
----------------------------------------------------------------------
Symbol table for: 'a.out' (x86_64)
----------------------------------------------------------------------
Index    n_strx   n_type             n_sect n_desc n_value
======== -------- ------------------ ------ ------ ----------------
[     0] 00000034 1e (PEXT SECT    ) 01     0000   0000000100003f30 '__Z6shrompi'
[     1] 00000040 1e (PEXT SECT    ) 01     0000   0000000100003f30 '__Z3fooi'
[     2] 00000049 1e (PEXT SECT    ) 01     0000   0000000100003f30 '__Z3bari'
[     3] 00000052 1e (PEXT SECT    ) 01     0080   0000000100003f30 '__Z4quuxi'
[     4] 0000005c 1e (PEXT SECT    ) 01     0000   0000000100003f40 '__Z3bazi'
[     5] 00000065 1e (PEXT SECT    ) 01     0080   0000000100003f40 '__Z5freepi'
[     6] 00000070 64 (N_SO         ) 00     0000   0000000000000000 '/Users/thakis/src/llvm-project/'
[     7] 00000090 64 (N_SO         ) 00     0000   0000000000000000 'bar.cc'
[     8] 00000097 66 (N_OSO        ) 03     0001   00000000624df143 '/Users/thakis/src/llvm-project/bar.o'
[     9] 00000001 2e (N_BNSYM      ) 01     0000   0000000100003ef0
[    10] 000000bc 24 (N_FUN        ) 01     0000   0000000100003ef0 '__Z1gi'
[    11] 00000001 24 (N_FUN        ) 00     0000   0000000000000040
[    12] 00000001 4e (N_ENSYM      ) 01     0000   0000000000000040
[    13] 00000001 2e (N_BNSYM      ) 01     0000   0000000100003f30
[    14] 000000c3 24 (N_FUN        ) 01     0000   0000000100003f30 '__Z4quuxi'
[    15] 00000001 24 (N_FUN        ) 00     0000   0000000000000010
[    16] 00000001 4e (N_ENSYM      ) 01     0000   0000000000000010
[    17] 00000001 2e (N_BNSYM      ) 01     0000   0000000100003f40
[    18] 000000cd 24 (N_FUN        ) 01     0000   0000000100003f40 '__Z5freepi'
[    19] 00000001 24 (N_FUN        ) 00     0000   0000000000000008
[    20] 00000001 4e (N_ENSYM      ) 01     0000   0000000000000008
[    21] 00000001 64 (N_SO         ) 01     0000   0000000000000000
[    22] 00000070 64 (N_SO         ) 00     0000   0000000000000000 '/Users/thakis/src/llvm-project/'
[    23] 000000d8 64 (N_SO         ) 00     0000   0000000000000000 'baz.cc'
[    24] 000000df 66 (N_OSO        ) 03     0001   00000000624df143 '/Users/thakis/src/llvm-project/baz.o'
[    25] 00000001 2e (N_BNSYM      ) 01     0000   0000000100003f50
[    26] 00000104 24 (N_FUN        ) 01     0000   0000000100003f50 '_main'
[    27] 00000001 24 (N_FUN        ) 00     0000   0000000000000050
[    28] 00000001 4e (N_ENSYM      ) 01     0000   0000000000000050
[    29] 00000001 64 (N_SO         ) 01     0000   0000000000000000
[    30] 00000002 0f (     SECT EXT) 01     0000   0000000100003ef0 '__Z1gi'
[    31] 00000009 0f (     SECT EXT) 01     0010   0000000100000000 '__mh_execute_header'
[    32] 0000001d 0f (     SECT EXT) 01     0000   0000000100003f50 '_main'
[    33] 00000023 01 (     UNDF EXT) 00     0100   0000000000000000 'dyld_stub_binder'
```

Note that ld64 emits all 8 functions to the symbol table (PEXT SECT or SECT EXT), but it doesn't emit N_FUN stabs entries for the functions that get ICF'd.

Now, lld:

```
% out/gn/bin/clang -fuse-ld=lld -O2 bar.o baz.o -Wl,--icf=all
% dsymutil -s a.out
----------------------------------------------------------------------
Symbol table for: 'a.out' (x86_64)
----------------------------------------------------------------------
Index    n_strx   n_type             n_sect n_desc n_value
======== -------- ------------------ ------ ------ ----------------
[     0] 00000070 64 (N_SO         ) 00     0000   0000000000000000 '/Users/thakis/src/llvm-project/bar.cc'
[     1] 00000096 66 (N_OSO        ) 03     0001   00000000624df143 '/Users/thakis/src/llvm-project/bar.o'
[     2] 000000bb 24 (N_FUN        ) 01     0000   00000001000003b0 '__Z4quuxi'
[     3] 00000001 24 (N_FUN        ) 00     0000   0000000000000010
[     4] 000000c5 24 (N_FUN        ) 01     0000   00000001000003b0 '__Z6shrompi'
[     5] 00000001 24 (N_FUN        ) 00     0000   0000000000000010
[     6] 000000d1 24 (N_FUN        ) 01     0000   00000001000003c0 '__Z5freepi'
[     7] 00000001 24 (N_FUN        ) 00     0000   0000000000000008
[     8] 000000dc 24 (N_FUN        ) 01     0000   00000001000003b0 '__Z3fooi'
[     9] 00000001 24 (N_FUN        ) 00     0000   0000000000000010
[    10] 000000e5 24 (N_FUN        ) 01     0000   00000001000003b0 '__Z3bari'
[    11] 00000001 24 (N_FUN        ) 00     0000   0000000000000010
[    12] 000000ee 24 (N_FUN        ) 01     0000   00000001000003c0 '__Z3bazi'
[    13] 00000001 24 (N_FUN        ) 00     0000   0000000000000008
[    14] 000000f7 24 (N_FUN        ) 01     0000   0000000100000370 '__Z1gi'
[    15] 00000001 24 (N_FUN        ) 00     0000   0000000000000040
[    16] 00000001 64 (N_SO         ) 01     0000   0000000000000000
[    17] 000000fe 64 (N_SO         ) 00     0000   0000000000000000 '/Users/thakis/src/llvm-project/baz.cc'
[    18] 00000124 66 (N_OSO        ) 03     0001   00000000624df143 '/Users/thakis/src/llvm-project/baz.o'
[    19] 00000149 24 (N_FUN        ) 01     0000   00000001000003d0 '_main'
[    20] 00000001 24 (N_FUN        ) 00     0000   0000000000000050
[    21] 00000001 64 (N_SO         ) 01     0000   0000000000000000
[    22] 0000000f 1e (PEXT SECT    ) 01     0000   00000001000003b0 '__Z4quuxi'
[    23] 00000019 1e (PEXT SECT    ) 01     0000   00000001000003b0 '__Z6shrompi'
[    24] 00000025 1e (PEXT SECT    ) 01     0000   00000001000003c0 '__Z5freepi'
[    25] 00000030 1e (PEXT SECT    ) 01     0000   00000001000003b0 '__Z3fooi'
[    26] 00000039 1e (PEXT SECT    ) 01     0000   00000001000003b0 '__Z3bari'
[    27] 00000042 1e (PEXT SECT    ) 01     0000   00000001000003c0 '__Z3bazi'
[    28] 00000002 0f (     SECT EXT) 01     0000   00000001000003d0 '_main'
[    29] 00000008 0f (     SECT EXT) 01     0000   0000000100000370 '__Z1gi'
[    30] 0000005c 0f (     SECT EXT) 01     0010   0000000100000000 '__mh_execute_header'
[    31] 0000004b 01 (     UNDF EXT) 00     0100   0000000000000000 'dyld_stub_binder'
```

lld emits the symbol table entries, but it also emits N_FUN entries for the folded functions, and it claims that they are in the .o file that the symbol they're folded into are in.

This confuses dsymutil:

```
% dsymutil a.out
warning: (x86_64)  could not find object file symbol for symbol __Z3fooi
warning: (x86_64)  could not find object file symbol for symbol __Z3bari
warning: (x86_64)  could not find object file symbol for symbol __Z3bazi
```

(See description of https://reviews.llvm.org/D123218 for many words on why exactly, but in summary it believes the N_FUN STABS entries and is then confused when the .o file doesn't define the symbols that the STABS info promises.)

**Comment by nico** [comment]
fix: https://reviews.llvm.org/D123252

**Comment by int3** [comment]
I think you meant s/omits/emits/ in the title :)

**Comment by nico** [comment]
Err, fixed. It's not like it means the epposite with that typo 🤦‍♂️

**Comment by nico** [comment]
Fixed in 8c1ea1ab815f84b6
